Sources/OpenAPIRuntime/EventStreams/ServerSentEventsDecoding.swift

```diff
@@ -0,0 +1,451 @@
+                    state = .finished
+                    return .returnNil
+                }
+            case .finished, .mutating: preconditionFailure("Invalid state")
+            }
+        }
+    }
+}
+
+/// A sequence that parses arbitrary byte chunks into lines using the Server-sent Events format.
+public struct ServerSentEventsLineDeserializationSequence<Upstream: AsyncSequence & Sendable>: Sendable
+where Upstream.Element == ArraySlice<UInt8> {
+
+    /// The upstream sequence.
+    private let upstream: Upstream
+
+    /// Creates a new sequence.
+    /// - Parameter upstream: The upstream sequence of arbitrary byte chunks.
+    public init(upstream: Upstream) { self.upstream = upstream }
+}
+
+extension ServerSentEventsLineDeserializationSequence: AsyncSequence {
+
+    /// The type of element produced by this asynchronous sequence.
+    public typealias Element = ArraySlice<UInt8>
+
+    /// The iterator of `ServerSentEventsLineDeserializationSequence`.
+    public struct Iterator<UpstreamIterator: AsyncIteratorProtocol>: AsyncIteratorProtocol
+    where UpstreamIterator.Element == Element {
+
+        /// The upstream iterator of arbitrary byte chunks.
+        var upstream: UpstreamIterator
+
+        /// The state machine of the iterator.
+        var stateMachine: StateMachine = .init()
+
+        /// Asynchronously advances to the next element and returns it, or ends the
+        /// sequence if there is no next element.
+        public mutating func next() async throws -> ArraySlice<UInt8>? {
+            while true {
+                switch stateMachine.next() {
+                case .returnNil: return nil
+                case .emitLine(let line): return line
+                case .noop: continue
+                case .needsMore:
+                    let value = try await upstream.next()
+                    switch stateMachine.receivedValue(value) {
+                    case .returnNil: return nil
+                    case .emitLine(let line): return line
+                    case .noop: continue
+                    }
+                }
+            }
+        }
+    }
+
+    /// Creates the asynchronous iterator that produces elements of this
+    /// asynchronous sequence.
+    public func makeAsyncIterator() -> Iterator<Upstream.AsyncIterator> {
+        Iterator(upstream: upstream.makeAsyncIterator())
+    }
+}
+
+extension ServerSentEventsLineDeserializationSequence.Iterator {
+
+    /// A state machine for parsing lines in Server-Sent Events.
+    ///
+    /// https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream
+    ///
+    /// This is not trivial to do with a streaming parser, as the end of line can be:
+    /// - LF
+    /// - CR
+    /// - CRLF
+    ///
+    /// So when we get CR, but have no more data, we want to be able to emit the previous line,
+    /// however we need to discard a LF if one comes.
```

Comment:
I think that we might be over reaching here and we can simplify even if it means delays. I.e. we can wait for the LF before emitting. It's covered explicitly by the spec:

> Since connections established to remote servers for such resources are expected to be long-lived, UAs should ensure that appropriate buffering is used. In particular, while line buffering with lines are defined to end with a single U+000A LINE FEED (LF) character is safe, block buffering or line buffering with different expected line endings can cause delays in event dispatch.

I read this as: if we can simplify here, we should. It's on the user to use good line endings.

Reply:
Are you proposing we don't support `<CR>`-delimited SSE at all? If so, that'd be the result - if we wait for a `<LF>` or end of stream before dispatching an event. You'd get all the events only after the server closed the connection at once, which basically means an unusable SSE stream.

Reply:
No, I was proposing that it's justified for us to simplify our logic such that the events are only eagerly emitted for LF-delimited streams as that was how I read this warning in the spec. Specifically, I think our logic is _correct_ but doing more than strictly necessary. If, given this understanding, we don't think this additional complexity is something we're desperate to shed then let's just keep it the way it is in this PR.